python/tvm/ir/instrument.py

```diff
@@ -0,0 +1,154 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+# pylint: disable=invalid-name,unused-argument
+"""Common pass instrumentation across IR variants."""
+import inspect
+import functools
+
+import tvm._ffi
+import tvm.runtime
+
+from . import _ffi_instrument_api
+
+
+@tvm._ffi.register_object("instrument.PassInstrument")
+class PassInstrument(tvm.runtime.Object):
+    """A pass instrument implementation.
+
+    Users don't need to interact with this class directly.
+    Instead, a `PassInstrument` instance should be created through `pass_instrument`.
+
+    See Also
+    --------
+    `pass_instrument`
+    """
+
+
+def _wrap_class_pass_instrument(pi_cls):
+    """Wrap a python class as pass instrument"""
+
+    class PyPassInstrument(PassInstrument):
+        """Internal wrapper class to create a class instance."""
+
+        def __init__(self, *args, **kwargs):
+            # initialize handle in cass pi_cls creation failed.fg
+            self.handle = None
+            inst = pi_cls(*args, **kwargs)
+
+            # check method declartion within class, if found, wrap it.
+            def create_method(method):
+                if hasattr(inst, method) and inspect.ismethod(getattr(inst, method)):
+
+                    def func(*args):
+                        return getattr(inst, method)(*args)
+
+                    func.__name__ = "_" + method
+                    return func
+                return None
+
+            # create runtime pass instrument object
+            # reister instance's run_before_pass, run_after_pass, set_up and tear_down method
+            # to it if present.
+            self.__init_handle_by_constructor__(
```

Comment:
due to an issue with how pytest interacts with Node classes, it's preferable that this is the first thing in `__init__`. if anything between line 47 and here raises an exception, pytest will cause an infinite recursion rather than printing the actual error. the constraint is that `self.handle` needs to be initialized.

Reply:
Yes, self.handle is initialized at line 49.
This wrapper function mimics ```_wrap_class_module_pass```  [transform.py:188](https://github.com/apache/tvm/blob/8934d6b5af1defc6ccd9e18fc26dac1a27f4546b/python/tvm/ir/transform.py#L188)